src/chai/ManagedArray.inl

```diff
@@ -133,47 +149,74 @@ template<typename T>
+  m_offset(0),
+  m_pointer_record(pointer_record),
+  m_is_slice(false)
 {
 }
 
+template<typename T>
+CHAI_INLINE
+CHAI_HOST ManagedArray<T> ManagedArray<T>::slice(size_t offset, size_t elems) {
+  ManagedArray<T> slice;
+  slice.m_resource_manager = m_resource_manager;
+  if(offset + elems > size()) {
+    CHAI_LOG("ManagedArray", "Invalid slice. No active pointer or index out of bounds");
+  } else {
+    slice.m_pointer_record = m_pointer_record;
+    slice.m_active_base_pointer = m_active_base_pointer;
+    slice.m_offset = offset + m_offset;
+    slice.m_active_pointer = m_active_base_pointer + slice.m_offset;
+    slice.m_elems = elems;
+    slice.m_is_slice = true;
+  }
+  return slice;
+}
+
 template<typename T>
 CHAI_INLINE
 CHAI_HOST void ManagedArray<T>::allocate(size_t elems, ExecutionSpace space, UserCallback const &cback) {
-  CHAI_LOG("ManagedArray", "Allocating array of size " << elems << " in space " << space);
+  if(!m_is_slice) {
+    CHAI_LOG("ManagedArray", "Allocating array of size " << elems << " in space " << space);
 
-  if (space == NONE) {
-    space = m_resource_manager->getDefaultAllocationSpace();
-  }
+    if (space == NONE) {
+      space = m_resource_manager->getDefaultAllocationSpace();
+    }
 
-  m_elems = elems;
-  m_pointer_record = m_resource_manager->allocate<T>(elems, space, cback);
-  m_active_pointer = static_cast<T*>(m_pointer_record->m_pointers[space]);
+    m_elems = elems;
+    m_pointer_record = m_resource_manager->allocate<T>(elems, space, cback);
+    m_active_base_pointer = static_cast<T*>(m_pointer_record->m_pointers[space]);
+    m_active_pointer = m_active_base_pointer; // Cannot be a slice
 
-  CHAI_LOG("ManagedArray", "m_active_ptr allocated at address: " << m_active_pointer);
+    CHAI_LOG("ManagedArray", "m_active_ptr allocated at address: " << m_active_pointer);
+  }
 }
 
 template<typename T>
 CHAI_INLINE
 CHAI_HOST void ManagedArray<T>::reallocate(size_t elems)
 {
-  CHAI_LOG("ManagedArray", "Reallocating array of size " << m_elems << " with new size" << elems);
+  if(!m_is_slice) {
+    CHAI_LOG("ManagedArray", "Reallocating array of size " << m_elems << " with new size" << elems);
 
-  m_elems = elems;
-  m_active_pointer =
-    static_cast<T*>(m_resource_manager->reallocate<T>(m_active_pointer, elems,
+    m_elems = elems;
+    m_active_base_pointer =
+      static_cast<T*>(m_resource_manager->reallocate<T>(m_active_base_pointer, elems,
                                                       m_pointer_record));
+    m_active_pointer = m_active_base_pointer; // Cannot be a slice
 
-  CHAI_LOG("ManagedArray", "m_active_ptr reallocated at address: " << m_active_pointer);
+    CHAI_LOG("ManagedArray", "m_active_ptr reallocated at address: " << m_active_pointer);
+  }
 }
 
 template<typename T>
 CHAI_INLINE
 CHAI_HOST void ManagedArray<T>::free()
 {
-  m_resource_manager->free(m_pointer_record);
+  if(!m_is_slice) m_resource_manager->free(m_pointer_record);
```

Comment:
It's probably worth doing a LOG message if someone attempts to free a slice - it's good that it's not  free'd here, but it's probably indicative of a bug in the main program.